src/runtime/naming/name.cpp

```diff
@@ -421,20 +422,31 @@ namespace hpx { namespace naming
         ///////////////////////////////////////////////////////////////////////
         std::int64_t replenish_credits(gid_type& gid)
         {
+            typedef std::unique_lock<gid_type::mutex_type> scoped_lock;
+            scoped_lock l(gid);
+            return replenish_credits_locked(l, gid);
+        }
+
+        HPX_EXPORT std::int64_t replenish_credits_locked(
+            std::unique_lock<gid_type::mutex_type>& l, gid_type& gid)
+        {
             std::int64_t added_credit = 0;
 
-            {
-                typedef std::unique_lock<gid_type::mutex_type> scoped_lock;
-                scoped_lock l(gid);
-                HPX_ASSERT(0 == get_credit_from_gid(gid));
+            HPX_ASSERT(0 == get_credit_from_gid(gid));
 
-                added_credit = naming::detail::fill_credit_for_gid(gid);
-                naming::detail::set_credit_split_mask_for_gid(gid);
-            }
+            added_credit = naming::detail::fill_credit_for_gid(gid);
+            naming::detail::set_credit_split_mask_for_gid(gid);
 
             gid_type unlocked_gid = gid;        // strips lock-bit
 
-            return agas::incref(unlocked_gid, added_credit);
+            std::int64_t result = 0;
+            {
+                hpx::util::scoped_unlock<std::unique_lock<gid_type::mutex_type> >
+                    ul(l);
+                 result = agas::incref(unlocked_gid, added_credit);
+            }
+
+            return result;
```

Comment:
Am 10.06.2016 4:14 nachm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>
> In src/runtime/naming/name.cpp:
>
> >
> >              gid_type unlocked_gid = gid;        // strips lock-bit
> >
> > -            return agas::incref(unlocked_gid, added_credit);
> > +            std::int64_t result = 0;
> > +            {
> > +
hpx::util::scoped_unlock<std::unique_lock<gid_type::mutex_type> >
> > +                    ul(l);
> > +                 result = agas::incref(unlocked_gid, added_credit);
> > +            }
> > +
> > +            return result;
>
> What is the rationale for this set of changes? Is the function
replenish_credits_locked() additioanlly called from elsewhere? From what I
can see, the only effect of the change would be an additonal lock/unlock on
the mutex.

One problem I discovered was that the promise gid never had credits. I
changed it such that it replenishes the credit at the first call. We need
to hold a lock on the gid while checking the credit count and replenishing,
that's why I introduced that function.

>
> Unlocking the mutex for agas::incref() is a valid change, though. However
there is no reson to re-lock the mutex afterwards.

I agree, it's not optimal, it all depends on what the caller expects. I
think it makes sense that a caller expects that his lock is still locked
after the function returns.

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.